### PR TITLE
athena audit logs - add migration script

### PR DESCRIPTION
**Pull request by tobiaszheller**
Part of https://github.com/gravitational/teleport.e/issues/894
RFD: https://github.com/gravitational/teleport/blob/master/rfd/0118-scalable-audit-logs.md

This PR adds migration script to migrate from dynamo to Athena logger.

More information about usage in README.md

### CODE REVIEW

File: lib/events/athena/dynamomigration/cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/events/athena/dynamomigration"
+)
+
+func main() {
+	timeStr := flag.String("exportTime", "", "exportTime is time in the past from which to export table data, empty for time.Now")
+	freshnessWindowStr := flag.String("freshnessWindow", "", "freshnessWindow defines duration if already ongoing or completed exports can be used, format time.Duration")
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/cmd/main.go]: https://github.com/gravitational/teleport/pull/27099/commits/e859f07c8c64a370e713d7b3f95011441c962a57
i have rework description of freshness windows in that commit, let me know if it's easier to understand now.

File: lib/events/athena/dynamomigration/migration_test.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		"testdata/dataObj2.json.gz": `{ "Item": { "EventIndex": { "N": "2147483647" }, "SessionID": { "S": "35f35254-92f9-46a2-9b05-8c13f712389b" }, "CreatedAtDate": { "S": "2023-05-22" }, "FieldsMap": { "M": { "cluster_name": { "S": "test.example.local" }, "uid": { "S": "46c03b4f-4ef5-4d86-80aa-4b53c7efc28f" }, "code": { "S": "T2005I" }, "ei": { "N": "2147483647" }, "time": { "S": "2023-05-22T12:12:21.966Z" }, "event": { "S": "session.upload" }, "sid": { "S": "35f35254-92f9-46a2-9b05-8c13f712389b" } } }, "EventType": { "S": "session.upload" }, "EventNamespace": { "S": "default" }, "CreatedAt": { "N": "1684757541" } } }`,
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/migration_test.go:45]: Problem is that those are json lines files, and vs code show warning that it's not valid json file :( But I can add them as files if that's important

**Comment by tobiaszheller** [comment]
@probakowski @xacrimon please take a look at review, sorry about size, but 380 lines from (dynamo_deserializer.go) is just copied from AWS sdk.

File: lib/events/athena/dynamomigration/README.md
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/README.md:1]: i will move it before merging

**Review by xacrimon — APPROVED** [review]
lgtm from a cursory overview but I lack domain-specific knawledge.

**Review by xacrimon — APPROVED** [review]
lgtm from a cursory overview but I lack domain-specific knowledge.

File: lib/events/athena/dynamomigration/cmd/main.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/events/athena/dynamomigration"
+)
+
+func main() {
+	timeStr := flag.String("exportTime", "", "exportTime is time in the past from which to export table data, empty for the current time")
+	freshnessWindowStr := flag.String("freshnessWindow", "", "freshnessWindow allows to use ongoing/completed exports instead of triggering new one, using for example '24h' will check if there is any export in exportTime-24h")
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/cmd/main.go]: reworked to use `exportARN` in https://github.com/gravitational/teleport/pull/27099/commits/8af0bf5a904cc094939cb15177324af8a6e4b192, it's way simpler, thanks for suggestion

File: lib/events/athena/dynamomigration/migration.go [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+	exportOutput, err := t.dynamoClient.ExportTableToPointInTime(ctx, &dynamodb.ExportTableToPointInTimeInput{
+		S3Bucket:     aws.String(t.Bucket),
+		TableArn:     aws.String(t.DynamoTableARN),
+		ExportFormat: dynamoTypes.ExportFormatDynamodbJson,
+		ExportTime:   aws.Time(t.ExportTime),
+		S3Prefix:     aws.String(t.Prefix),
+	})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/migration.go:285]: I have added handling of failedStatus in https://github.com/gravitational/teleport/pull/27099/commits/46e2537603fdebac8ac5a8cd41c7332f06971fa7, it's checked in `waitForCompletedExport`

File: lib/events/athena/dynamomigration/migration.go [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+	checkpointsPerWorker := map[int]string{}
+
+	// write information about last checkpoint which can be used to continue.
+	workerCtx, workerCancel := context.WithCancel(ctx)
+	defer workerCancel()
+
+	errorC := make(chan error, t.NoOfEmitWorkers)
+
+	var wg sync.WaitGroup
+	wg.Add(t.NoOfEmitWorkers)
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/migration.go]: yeh, good call, reworked in https://github.com/gravitational/teleport/pull/27099/commits/a848906fbb5e17ae4b463ba5ad03962fb6525fd3

**Comment by tobiaszheller** [comment]
@rosstimothy it should be ready for re-review

**Review by probakowski — APPROVED** [review]

File: lib/events/athena/dynamomigration/migration.go [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+		return nil, trace.Wrap(err)
+	}
+
+	event, err := events.FromEventFields(eventFields)
+	return event, trace.Wrap(err)
+}
+
+func (t *task) downloadFromS3(ctx context.Context, key string) (*os.File, error) {
+	originalName := path.Base(key)
+	path := path.Join(os.TempDir(), originalName)
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/migration.go]: > Who/What cleans up this directory? The OS?
yep, OS

> What happens if I run this in a kubernetes pod that has a read-only filesystem?
I will add it to requirements to use writeable filesystem.

> Should we add a flag that allows the user dictate where these should be saved?
added in https://github.com/gravitational/teleport/pull/27099/commits/2159e9230f96018491392497f2f79260e39f6af6

File: lib/events/athena/dynamomigration/README.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
+# Dynamomigration tool
+
+Dynamomigration tool allows to export Teleport audit events logs from DynamoDB
+table into Athena Audit log.
+It's using DynamoDB export to S3 to export data.
+
+Requirements:
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/README.md:7]: added in https://github.com/gravitational/teleport/pull/27099/commits/1f88c08f952313562887c017cd0b2fc3e613b17a

File: lib/events/athena/dynamomigration/migration_test.go [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+	mt := &task{
+		s3Downloader: &fakeDownloader{
+			dataObjects: testDataObjects,
+		},
+		eventsEmitter: emitter,
+		Config: Config{
+			Logger:          utils.NewLoggerForTests(),
+			NoOfEmitWorkers: 5,
+			bufferSize:      10,
+			CheckpointPath:  path.Join(os.TempDir(), "migration-tests.json"),
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/migration_test.go]: https://github.com/gravitational/teleport/pull/27099/commits/74ed971b2f06c0da05efd25ccc3827105fa6571b

File: lib/events/athena/dynamomigration/README.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+It is recommended to test export first using `-dryRun` flag. DryRun does not emit any events,
+it makes sure that export is in valid format and events can be parsed.
+
+Dry run example:
+
+```shell
+./dynamomigration -dynamoARN='arn:aws:dynamodb:region:account:table/tablename' \
+  -exportPath='s3://bucket/prefix' \
+  -freshnessWindow=24h \
```
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/README.md]: fixed in https://github.com/gravitational/teleport/pull/27099/commits/5ba2d29742d7ac4dd2c3837ac38c95072e4f0314

**Review by rosstimothy — APPROVED** [review]

File: lib/events/athena/dynamomigration/README.md
[the remark this answers is not in the document]
Comment (tobiaszheller) [reply lib/events/athena/dynamomigration/README.md:1]: done, moved to examples

**Review by jimbishopp — APPROVED** [review]

**Pull request merged**